src/irmin-pack/inode.ml

```diff
@@ -506,6 +506,95 @@ struct
+
+      type 'a pointer = { index : int; pointer : hash; tree : 'a }
+      [@@deriving irmin]
+
+      type 'a tree = { depth : int; length : int; pointers : 'a pointer list }
+      [@@deriving irmin]
+
+      type t = Tree of t tree | Value of entry list [@@deriving irmin]
+
+      let to_entry (name, v) =
+        match v with
+        | `Contents (hash, m) -> { name; kind = Contents m; hash }
+        | `Node hash -> { name; kind = Node; hash }
+
+      let of_entry e =
+        ( e.name,
+          match e.kind with
+          | Contents m -> `Contents (e.hash, m)
+          | Node -> `Node e.hash )
+
+      type error = [ `Integrity_error of hash * hash * t ]
+
+      let pp = Irmin.Type.pp_json t
+
+      let pp_error ppf (`Integrity_error (got, expected, t)) =
+        Fmt.pf ppf
+          "integrity error: %a has hash %a, while hash %a was expected." pp t
+          pp_hash got pp_hash expected
+    end
+
+    let to_concrete (la : 'ptr layout) (t : 'ptr t) =
+      let rec aux t =
+        match t.v with
+        | Tree tr ->
+            ( Lazy.force t.hash,
+              Concrete.Tree
+                {
+                  depth = tr.depth;
+                  length = tr.length;
+                  pointers =
+                    Array.fold_right
+                      (fun e (i, acc) ->
+                        match e with
+                        | None -> (i + 1, acc)
+                        | Some t ->
+                            let pointer, tree = aux (Ptr.target la t) in
+                            (i + 1, { Concrete.index = i; tree; pointer } :: acc))
+                      tr.entries (0, [])
+                    |> snd;
+                } )
+        | Values l ->
+            ( Lazy.force t.hash,
+              Concrete.Value (List.map Concrete.to_entry (StepMap.bindings l))
+            )
+      in
+      snd (aux t)
+
+    exception Invalid_hash of hash * hash * Concrete.t
+
+    let hash_equal = Irmin.Type.(unstage (equal hash_t))
+
+    let of_concrete t =
+      let hash v = Bin.V.hash (to_bin_v Total v) in
+      let rec aux t =
+        match t with
+        | Concrete.Value l ->
+            Values (StepMap.of_list (List.map Concrete.of_entry l))
+        | Concrete.Tree tr ->
+            let entries = Array.make Conf.entries None in
+            List.iter
+              (fun { Concrete.index; pointer; tree } ->
+                let v = aux tree in
+                let hash = hash v in
+                if not (hash_equal hash pointer) then
+                  raise (Invalid_hash (hash, pointer, t));
+                let t = { hash = lazy pointer; stable = false; v } in
+                entries.(index) <- Some (Ptr.of_target Total t))
+              tr.pointers;
+            Tree { depth = tr.depth; length = tr.length; entries }
+      in
+      try
+        let v = aux t in
+        let hash = hash v in
+        Ok { hash = lazy hash; stable = false; v }
```

Comment:
Always using `stable = false` here is incorrect. It depends on the status of `t` in `of_concrete t`:
- If `t` is a Tree with `depth = 0`, it should stabilized.
- If `t` is a Tree with `depth > 0`, it's okay.
- If `t` is the empty Values, it should be stable.
- If `t` is a non-empty Value, we don't know (see `is_root` below). 

I see a few solutions:
1. Add a `stable` boolean at each level of `Concrete.t`
2. Add a `depth` integer in `Concrete.Values`.
3. Forbid the use of `of_concrete` on non-root inodes
4. Add `stabilize:bool` parameter

Reply:
We should call `integrity_check` on the output too. Checking the hash of the pointers may not be enough to catch all the integrity problems.